http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/SocketOptionsConfiguration.java

```diff
@@ -0,0 +1,215 @@
+        return domain;
+    }
+
+    /**
+     * @return socket type
+     */
+    public SocketOptions.SocketType type() {
+        return type;
+    }
+
+    /**
+     * @return number of milliseconds before a connection will be considered timed out
+     */
+    public int connectTimeoutMs() {
+        return connectTimeoutMs;
+    }
+
+    /**
+     * @return number of seconds between TCP keepalive packets being sent to the peer
+     */
+    public int keepAliveIntervalSecs() {
+        return keepAliveIntervalSecs;
+    }
+
+    /**
+     * @return number of seconds to wait for a keepalive response before considering the connection timed out
+     */
+    public int keepAliveTimeoutSecs() {
+        return keepAliveTimeoutSecs;
+    }
+
+    public static Builder builder() {
+        return new DefaultSocketOptionsConfigurationBuilder();
+    }
+
+    /**
+     * A builder for {@link SocketOptionsConfiguration}.
+     *
+     * <p>All implementations of this interface are mutable and not thread safe.</p>
+     */
+    public interface Builder {
+
+        /**
+         * Sets the socket domain
+         * @param domain socket domain
+         * @return Builder
+         */
+        Builder domain(SocketOptions.SocketDomain domain);
+
+        /**
+         * Sets the socket type
+         * @param type socket type
+         * @return Builder
+         */
+        Builder type(SocketOptions.SocketType type);
+
+        /**
+         * Sets the number of milliseconds before a connection will be considered timed out
+         * @param connectTimeoutMs number of milliseconds before a connection will be considered timed out
+         * @return Builder
+         */
+        Builder connectTimeoutMs(int connectTimeoutMs);
+
+        /**
+         * Sets the number of seconds between TCP keepalive packets being sent to the peer
+         * @param keepAliveIntervalSecs number of seconds between TCP keepalive packets being sent to the peer
+         * @return Builder
+         */
+        Builder keepAliveIntervalSecs(int keepAliveIntervalSecs);
+
+        /**
+         * Sets the number of seconds to wait for a keepalive response before considering the connection timed out
+         * @param keepAliveTimeoutSecs number of seconds to wait for a keepalive response before considering the connection timed out
+         * @return Builder
+         */
+        Builder keepAliveTimeoutSecs(int keepAliveTimeoutSecs);
+
+        SocketOptionsConfiguration build();
+    }
+
+    /**
+     * An SDK-internal implementation of {@link Builder}.
+     */
+    private static final class DefaultSocketOptionsConfigurationBuilder implements Builder {
+
+        private SocketOptions.SocketDomain domain = SocketOptions.SocketDomain.IPv6;
+        private SocketOptions.SocketType type = SocketOptions.SocketType.STREAM;
+        private int connectTimeoutMs = 3000;
+        private int keepAliveIntervalSecs = 0;
+        private int keepAliveTimeoutSecs = 0;
```

Comment:
Instead of defining defaults here, I think we should put the configured values into the standard options map like below. The SDK will use the global defaults automatically for any options that are not configured 

https://github.com/aws/aws-sdk-java-v2/blob/b1f89898a7cd83626c0b1a01bd671594d14cd322/http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/NettyNioAsyncHttpClient.java#L542-L545

https://github.com/aws/aws-sdk-java-v2/blob/b1f89898a7cd83626c0b1a01bd671594d14cd322/http-client-spi/src/main/java/software/amazon/awssdk/http/SdkHttpConfigurationOption.java#L155-L171

Reply:
Can move those there, sure.

Reply:
Based on the discussion in https://github.com/aws/aws-sdk-java-v2/pull/3457#discussion_r987166959, i think it doesn't make sense to list defaults for these